src/factories/createSqlTag.ts

```diff
@@ -92,19 +96,62 @@ const sql: SqlTaggedTemplate = (
     }
   }
 
-  const query: SqlTokenType = {
+  return {
     sql: rawSql,
-    type: SqlToken,
     values: parameterValues,
   };
+};
+
+const sql = (
+  parts: readonly string[],
+  ...args: readonly ValueExpression[]
+) => {
+  const {
+    sql: sqlText,
+    values,
+  } = createQuery(parts, args);
+
+  const query = {
+    sql: sqlText,
+    type: SqlToken,
+    values,
+  };
 
   Object.defineProperty(query, 'sql', {
     configurable: false,
     enumerable: true,
     writable: false,
   });
 
-  return query;
+  return query as unknown as SqlSqlToken<QueryResultRow>;
+};
+
+sql.type = <T extends ZodTypeAny>(zodObject: T) => {
+  return (
+    parts: readonly string[],
+    ...args: readonly ValueExpression[]
+  ) => {
+    const {
+      sql: sqlText,
+      values,
+    } = createQuery(parts, args);
+
+    const query = {
+      sql: sqlText,
+      type: SqlToken,
+      values,
+      // @ts-expect-error We know that `zodObject` is ZodObject
+      zodObject: zodObject.strict(),
+    };
+
+    Object.defineProperty(query, 'sql', {
+      configurable: false,
+      enumerable: true,
+      writable: false,
+    });
+
+    return query as unknown as SqlSqlToken<T>;
```

Comment:
Shouldn't this be something like 
```suggestion
    return query as unknown as SqlSqlToken<z.infer<T>>;
```

Or if you take the Parser<T> suggestion this line would be correct as is